File: ruby/021_ComeBackInOnePiece.rb
# 021 - Come Back in One Piece（★5）
# 強連結成分分解

N, M = gets.split.map(&:to_i)
GOTO = Array.new(N) { [] }
BACKTO = Array.new(N) { [] }
M.times do
  a, b = gets.split.map { _1.to_i - 1 }
  GOTO[a] << b
  BACKTO[b] << a
end

def dfs1(pos)
  return if @used[pos]
  @used[pos] = true
  GOTO[pos].each { dfs1(_1) }
  @log << pos
end

def dfs2(pos)
  return 0 if @used[pos]
  @used[pos] = true
  cnt = 1
  BACKTO[pos].each { cnt += dfs2(_1) }
  cnt
end

@used = [false] * N
@log = []
N.times { dfs1(_1) }

@used = [false] * N
ans = 0
ans = @log.reverse.sum do |pos|
  cnt = dfs2(pos)
  cnt * (cnt - 1) / 2
end

puts ans

File: kotlin/021_ComeBackInOnePiece.kt
/*
 * 021 - Come Back in One Piece（★5）
 * 強連結成分分解
 */

// 再帰関数をそのまま実行するとスタックオーバーフローしREとなる。
// それを回避するためThreadとして実行する必要があり、メイン処理をメソッド定義
fun solve() {
  val (N, M) = readLine()!!.split(" ").map(String::toInt)
  val GOTO = MutableList(N + 1) { mutableListOf<Int>() }
  val BACKTO = MutableList(N + 1) { mutableListOf<Int>() }

  for (i in 1 .. M) {
    val (a, b) = readLine()!!.split(" ").map(String::toInt)
    GOTO[a].add(b)
    BACKTO[b].add(a)
  }

  var used = MutableList(N + 1) { false }
  var log = mutableListOf<Int>()
  fun dfs1(pos: Int) {
    if (used[pos]) return
    used[pos] = true
    GOTO[pos].forEach(::dfs1)
    log.add(pos)
  }

  (1 .. N).forEach(::dfs1)

  used = MutableList(N + 1) { false }
  fun dfs2(pos: Int): Long {
    if (used[pos]) return 0
    used[pos] = true
    var cnt = 1L
    BACKTO[pos].forEach { cnt += dfs2(it) }
    return cnt
  }

  val ans = log.reversed().fold(0L) {acc, i ->
    val cnt = dfs2(i)
    acc + cnt * (cnt - 1) / 2
  }
  println(ans)
}

fun main() {
  Thread(null,::solve, "solve", 1.shl(26)).start()
}
